### ISSUE TITLE
Wrong normalization implementation

### ISSUE THREAD

**Issue by ug0x01** [CaioSBC/RLPortfolio#9]
Hello, I saw that you are using normal normalizers instead of rolling ones 

https://github.com/CaioSBC/RLPortfolio/blob/071cf75da9ddf95ec62210627b971086eec6f18f/rlportfolio/data/scalers.py#L63

Let me explain why it could cause getting different results in live market. I will use random numbers just to explain

```python
import random

#creating random prices for example
a = [250]

for _ in range(15):
    a.append(random.randint(-11, 11) + a[-1])

print(a)
```

That is the output that I got `[250, 258, 248, 248, 256, 260, 268, 259, 253, 259, 263, 264, 272, 279, 270, 269]`

Now let's see the difference

```python
from sklearn.preprocessing import MinMaxScaler
import numpy as np

b = np.array(a).reshape(-1, 1)
b = MinMaxScaler().fit_transform(b)

print(b)
```

Output;

```
[[0.06451613]
 [0.32258065]
 [0.        ]
 [0.        ]
 [0.25806452]
 [0.38709677]
 [0.64516129]
 [0.35483871]
 [0.16129032]
 [0.35483871]
 [0.48387097]
 [0.51612903]
 [0.77419355]
 [1.        ]
 [0.70967742]
 [0.67741935]]
```

```python
#case that we don't see the future just like the real life implementation

def check(data):
    data = np.array(data).reshape(-1, 1)
    e = []

    for i in range(1, len(data)):
        e.append(MinMaxScaler().fit_transform(data[0:i]))

    return e

e = check(a)
#what would it seem like at the 10th step if it were just doing the same normalization in real life
print(e[10]) 
```

Output:

```
[[0.1 ]
 [0.5 ]
 [0.  ]
 [0.  ]
 [0.4 ]
 [0.6 ]
 [1.  ]
 [0.55]
 [0.25]
 [0.55]
 [0.75]]
```

In training, let's assume the window size is 5, the input data for the same step will be

```
[[0.64516129]
 [0.35483871]
 [0.16129032]
 [0.35483871]
 [0.48387097]]
```

But in real life the input will be

```
[[1.  ]
 [0.55]
 [0.25]
 [0.55]
 [0.75]]
```

As you can see it would return much more different result in real life

Here's the basic manual rolling min max scaler to fix that issue:

```python
def RollingMinMaxScaler(data, window_size):
    temp = []

    for j in range(len(data) - window_size):
        tempytemp = []
        
        for i in range(window_size):
            tempytemp.append((data[j+i] - min(data[j:j+window_size])) / (max(data[j:j+window_size]) - min(data[j:j+window_size])))

        temp.append(tempytemp)

    return temp

b = RollingMinMaxScaler(a, 5)
print(b)
```

Output:

```
[[0.2, 1.0, 0.0, 0.0, 0.8], [0.8333333333333334, 0.0, 0.0, 0.6666666666666666, 1.0], [0.0, 0.0, 0.4, 0.6, 1.0], [0.0, 0.4, 0.6, 1.0, 0.55], [0.2, 0.4666666666666667, 1.0, 0.4, 0.0], [0.4666666666666667, 1.0, 0.4, 0.0, 0.4], [1.0, 0.4, 0.0, 0.4, 0.6666666666666666], [0.5454545454545454, 0.0, 0.5454545454545454, 0.9090909090909091, 1.0], [0.0, 0.3157894736842105, 0.5263157894736842, 0.5789473684210527, 1.0], [0.0, 0.2, 0.25, 0.65, 1.0], [0.0, 0.0625, 0.5625, 1.0, 0.4375]]
```

It will use more ram for the data but it's the only way to prevent this issue (also you will need to change data processing too)

Have a good day!

**Comment by ug0x01** [comment]
Or you can simply scale for every input instead of scaling it all at once like 

```python
df_portfolio = GroupByScaler(columns = ["close", "high", "low"], by="tic").fit_transform(df_portfolio)
```

**Comment by ug0x01** [comment]
Just scale your inputs by window size when you are going to use it

**Comment by ug0x01** [comment]
Generally speaking, the difference between the lowest and highest prices in the datasets are being too large (for example, Bitcoin increased almost 20x in the last 5 years) and because of that, nearly 80% of the training will occur under conditions and inputs that are significantly different from the real implementation

**Comment by CaioSBC** [comment]
Hello! 

Sorry for the late response, I was traveling these days and didn't access GitHub.

The groupby scaler is just a tool so that you can use a scikit-learn scaler individually by each feature. When I use it, I apply a `fit_transform` in the training data and then simply `fit` the test data. From my personal experience, models can be trained with such normalized data and generate a good policy of actions, but the optimization usually gets stuck in local maxima. It is, as you stated, the easiest way to normalize data but issues can appear, for example, if prices in the test data are too high and values considerably bigger than 1 are generated by the normalization (this issue is mitigated by using online learning during testing, but it still exists).

However, as you noticed, it is, theoretically speaking, better to scale the data by window size. This solution is currently present in the environment through the parameter `state_normalization` in its constructor. When using this normalization method, every time the environment returns an observation in the `step` or `reset` method, its values are normalized, so you don't need to normalize the input dataframe - the environment will handle everything for you. Usually, the academic literature normalizes the observation by dividing it by the last closing price of each asset, but there are a few other viable methods implemented as you can see [here](https://rlportfolio.readthedocs.io/en/latest/environment/portfolio_optimization_env.html#normalization-methods).

However, during my personal experiences, the usage of such state normalization leads to agents with considerably variable performance: sometimes, extremely good agents are generated, while many other times, the resulting agent underperforms. I still couldn't figure out the reason behind this behavior, so maybe you can help me out!

I appreciate your interest in this library. Any suggestions are welcome.

**Comment by ug0x01** [comment]
Hello @CaioSBC!

> When using this normalization method, every time the environment returns an observation in the step or reset method, its values are normalized, so you don't need to normalize the input dataframe

Ohh, I didn't notice that and yes I totally forgot that in RL envs, there are step and reset steps (at first I assumed that you are still just shifting the prepared data) and this steps should prevent the issue.

> However, during my personal experiences, the usage of such state normalization leads to agents with considerably variable performance: sometimes, extremely good agents are generated, while many other times, the resulting agent underperforms. I still couldn't figure out the reason behind this behavior, so maybe you can help me out!

So different scalers results in different performance scores. I assume it's about the network's topology itself. Every scaler classifies the data by it's different features. F.e min max scaler classifies the data by the rate of the value compared to other data and for the robust scaler, values are comperad to median data (more likely to figure time series features) and since neural nets are black boxes, we can't actually know what it's exactly learns and it's more suitable to which strengths. So I guess it's just another hyperparameter which should be tested.

Please right me if I missunderstood a part. I will try my best to help with the issue
